[FAB-18508] ledger utility always outputs txNum

shimos opened the pull request:
#### Type of change

- Bug fix

#### Description

This patch fixes a bug reported in FAB-18508, and makes the ledger troubleshooting utility output always the record "txNum".

#### Additional details

The bug occurred when the target transaction is the first transaction of a block (txNum == 0). This patch removes the `omitempty` attributes from all the members of a record.

Note that this patch also changes the output when a key is missing in either snapshot. Instead of an empty object ({}) for the missing snapshot, the entire key (either "snapshotrecord1" or "snapshotrecord2") will be omitted.
(Refer to the changes in the test code)

#### Related issues

https://jira.hyperledger.org/projects/FAB/issues/FAB-18508

shimos committed:
[FAB-18508] ledger utility always outputs txNum

This patch fixes a bug reported in FAB-18508, and makes the ledger
troubleshooting utility output always the record "txNum".

The bug occurred when the target transaction is the first transaction of
a block (i.e. txNum == 0). This patch removes the `omitempty` attributes
from all the members of a record.

Note that this patch also changes the output when a key is missing in
either snapshot. For a missing snapshot, 'null' will be output instead
of an empty object ({}).

Signed-off-by: Taku Shimosawa <taku.shimosawa.uf@hitachi.com>

manish-sethi commented on internal/ledger/compare_test.go:
```diff
@@ -498,13 +484,11 @@ func TestJSONArrayFileWriter(t *testing.T) {
 				"blockNum" : 472,
 				"txNum" : 61
-			},
-			"snapshotrecord2" : {}
```
This was an intentional entry that was chosen to represent a missing record. Unless there is a strong reason, I would prefer to keep it as is.

manish-sethi reviewed (commented):
Thanks @shimos for your contribution. Left you a comment.

shimos commented:
@manish-sethi Thank you for the review. I have revised the patch to keep the JSON structure.

manish-sethi commented on internal/ledger/compare.go:
```diff
@@ -186,24 +186,20 @@ func Compare(snapshotDir1 string, snapshotDir2 string, outputFile string) (count
+	Namespace string      `json:"namespace,omitempty"`
+	Key       string      `json:"key,omitempty"`
+	Record1   interface{} `json:"snapshotrecord1,omitempty"`
+	Record2   interface{} `json:"snapshotrecord2,omitempty"`
```
It was better to use the concrete type, as before. On a second thought, as an alternate we could remove "omitempty" in the json annotation for `Record1` and `Record2`. That will force to print a nil object as `null` in the final JSON -- which I feel would be even clearer than `{}`.

Sorry for an additional round of change on this.

shimos commented:
@manish-sethi ok, so I have reverted to the first patch, and deleted 'omitempty' from `snapshotrecord1` and `snapshotrecord2` so that it will produce `null` for the missing key rather than deleting the key itself.

manish-sethi commented:
Thanks @shimos for your contribution!
